index.rst

```diff
@@ -338,4 +338,121 @@ smaller than a single alert message.
+3+1 consumers, 5+1, and 10+1, testing both with postage stamps and without.
+
+Below shows the memory utilization of Kafka with an increased number of consumers and larger allotted
+instances on AWS.
+
+.. figure:: _static/memoryConsumers.png
+   :width: 55%
+   :align: center
+   :name: figure-14
+
+   Kafka container memory utilization against number of consumers.
+
+For the 1+1 consumer test, instances were only allotted 32 GB of memory.  The other tests had 256 GiB memory
+available and utilize more memory.  Tests with stamps included in alerts are significantly more memory
+intensive for Kafka than tests without including postage stamps cutouts.
+
+For these tests there is no difference for the network traffic into Kafka, as we are still sending 10,000 alerts
+in each visit burst.  The network traffic out of Kafka increases linearly with more consumers, as expected,
+up to 180000 KiBps (180 Mbps) for 10+1 consumers, since each consumer pulls its own stream.
+
+.. figure:: _static/netOutConsumers.png
+   :width: 55%
+   :align: center
+   :name: figure-15
+
+   Network traffic out of Kafka against number of consumers.
+
+Below shows the length of time it takes for producers to serialize alerts into Avro and produce to the Kafka
+broker.  Having more consumers has a small impact on the producer time.
+For the 1+1 consumer test, we used only one producer sending 10,000 alerts (unparallelized), which is
+why the time is slightly higher.  For the other three tests, the serialization and producing time increases
+slightly and linearly to 9 seconds.  With more producers producing in parallel (one per CCD),
+this number would likely be lower.
+
+.. figure:: _static/serialTimeConsumers.png
+   :width: 55%
+   :align: center
+   :name: figure-16
+
+   Serialization and Kafka submission time for producers against number of consumers.
+
+The transit time of alerts, the time it takes for alerts to be read by the consumers after they have been
+submitted into the Kafka queue, increases with number of consumers.  Below shows the relationship between
+the transit time and number of consumers.
+
+
+.. figure:: _static/transitTimeConsumers.png
+   :width: 55%
+   :align: center
+   :name: figure-17
+
+   Transit time of alerts to consumers against number of consumers.
+
+For the test with 10+1 consumers including postage stamp cutouts, the average transit time is very large.
+The consumers are unable to keep up with bursts of alerts every 39 seconds.  Below shows the transit time lag
+increasing for alert bursts for 300 visits.
+
+.. figure:: _static/slowConsumer.png
+   :width: 55%
+   :align: center
+   :name: figure-18
+
+   Length of time from Kafka to consumer receipt and processing in seconds (y axis) vs. visit number (x axis).
+
+Fortunately, even with the very slow consumer read time, the producer submission time does not appear to be
+affected.
+
+
+Thoughts and Recommendations
```

Comment:
This heading is at the same level as “Scaling Alert Consumers”, but, in fact, everything that you're discussing here relates *only* to the consumer tests; there are no thoughts or recommendations about the earlier material. Given that, it seems like it should probably be a subsection of the material on consumers.

Reply:
No, not really.  The first paragraph about stamps also applies to some of the producer scaling experiments.  The stamps especially slow the transit time with increased number of producers, more so than without stamps.